ampd/src/handlers/solana_verify_msg.rs

```diff
@@ -0,0 +1,383 @@
+
+type Result<T> = error_stack::Result<T, Error>;
+
+#[derive(Deserialize, Debug)]
+pub struct Message {
+    pub message_id: Base58SolanaTxSignatureAndEventIndex,
+    pub destination_address: String,
+    pub destination_chain: ChainName,
+    #[serde(deserialize_with = "crate::solana::deserialize_pubkey")]
+    pub source_address: Pubkey,
+    pub payload_hash: Hash,
+}
+
+#[derive(Deserialize, Debug)]
+#[try_from("wasm-messages_poll_started")]
+struct PollStartedEvent {
+    poll_id: PollId,
+    source_chain: ChainName,
+    expires_at: u64,
+    messages: Vec<Message>,
+    participants: Vec<TMAddress>,
+}
+
+pub struct Handler<C: SolanaRpcClientProxy> {
+    verifier: TMAddress,
+    voting_verifier_contract: TMAddress,
+    rpc_client: C,
+    latest_block_height: Receiver<u64>,
+    event_match_context: MatchContext,
+}
+
+impl<C: SolanaRpcClientProxy> Handler<C> {
+    pub fn new(
+        verifier: TMAddress,
+        voting_verifier_contract: TMAddress,
+        rpc_client: C,
+        latest_block_height: Receiver<u64>,
+    ) -> Self {
+        let event_match_context = MatchContext::new(&axelar_solana_gateway::ID.to_string());
+
+        Self {
+            verifier,
+            voting_verifier_contract,
+            rpc_client,
+            event_match_context,
+            latest_block_height,
+        }
+    }
+
+    fn vote_msg(&self, poll_id: PollId, votes: Vec<Vote>) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.as_ref().clone(),
+            contract: self.voting_verifier_contract.as_ref().clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::Vote { poll_id, votes })
+                .expect("vote msg should serialize"),
+            funds: vec![],
+        }
+    }
+
+    async fn fetch_message(
+        &self,
+        msg: &Message,
+    ) -> Option<(solana_sdk::signature::Signature, UiTransactionStatusMeta)> {
+        let signature = solana_sdk::signature::Signature::from(msg.message_id.raw_signature);
+        self.rpc_client
+            .get_tx(&signature)
+            .await
+            .map(|tx| (signature, tx))
+    }
+}
+
+#[async_trait]
+impl<C: SolanaRpcClientProxy> EventHandler for Handler<C> {
+    type Err = Error;
+
+    async fn handle(&self, event: &events::Event) -> Result<Vec<Any>> {
+        if !event.is_from_contract(self.voting_verifier_contract.as_ref()) {
+            return Ok(vec![]);
+        }
+
+        let PollStartedEvent {
+            poll_id,
+            source_chain,
+            messages,
+            expires_at,
+            participants,
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(vec![])
+            }
+            event => event.change_context(DeserializeEvent)?,
+        };
+        if !participants.contains(&self.verifier) {
+            return Ok(vec![]);
+        }
+
```

Comment:
should check the source chain

Reply:
I am assuming :tm: (correct me if am wrong) each chain deploys it's own voting verifier cosmwasm contract. If that's true, wouldn't [this check](https://github.com/eigerco/axelar-amplifier/blob/00c7f11e1bdb7df044e9267518b87ca42e056206/ampd/src/handlers/solana_verify_msg.rs#L101-L103) be enough ?

Reply:
We don't see the source chain check in any implementation apart from evm

Reply:
Added the paramter and tests at 69e0742252f575d40c84ceab02a4ee34bf4be56f.

Also ran cargo fmt at e6a5f4c67d365aea51b5545d7db373db9eb62c28